detection_rules/rule_loader.py

```diff
@@ -4,132 +4,203 @@
-    file_lookup = file_lookup or load_rule_files(verbose=verbose)
-
-    failed = False
-    rules: List[TOMLRule] = []
-    errors = []
-    queries = []
-    query_check_index = []
-    rule_ids = set()
-    rule_names = set()
-
-    for rule_file, rule_contents in file_lookup.items():
+
+def metadata_filter(**metadata) -> Callable[[TOMLRule], bool]:
+    """Get a filter callback based off rule metadata"""
+    flt = dict_filter(metadata)
+
+    def callback(rule: TOMLRule) -> bool:
+        target_dict = rule.contents.metadata.to_dict()
+        return flt(target_dict)
+
+    return callback
+
+
+production_filter = metadata_filter(maturity="production")
+deprecate_filter = metadata_filter(maturity="deprecated")
+
+
+class RuleCollection:
+    """Collection of rule objects."""
+
+    __default = None
+
+    def __init__(self, rules: Optional[List[TOMLRule]] = None):
+        self.id_map: Dict[definitions.UUIDString, TOMLRule] = {}
+        self.file_map: Dict[Path, TOMLRule] = {}
+        self.rules: List[TOMLRule] = []
+        self.frozen = False
+
+        self._toml_load_cache: Dict[Path, dict] = {}
+
+        for rule in (rules or []):
+            self.add_rule(rule)
+
+    def __len__(self):
+        """Get the total amount of rules in the collection."""
+        return len(self.rules)
+
+    def __iter__(self):
+        """Iterate over all rules in the collection."""
+        return iter(self.rules)
+
+    def __contains__(self, rule: TOMLRule):
+        """Check if a rule is in the map by comparing IDs."""
+        return rule.id in self.id_map
+
+    def filter(self, cb: Callable[[TOMLRule], bool]) -> 'RuleCollection':
+        """Retrieve a filtered collection of rules."""
+        filtered_collection = RuleCollection()
+
+        for rule in filter(cb, self.rules):
+            filtered_collection.add_rule(rule)
+
+        return filtered_collection
+
+    def add_rule(self, rule: TOMLRule):
+        assert not self.frozen, f"Unable to add rule {rule.name} {rule.id} to a frozen collection"
+        assert rule.id not in self.id_map, \
+            f"Rule ID {rule.id} for {rule.name} collides with rule {self.id_map.get(rule.id).name}"
+
+        if rule.path is not None:
+            rule.path = rule.path.resolve()
+            assert rule.path not in self.file_map, f"Rule file {rule.path} already loaded"
+            self.file_map[rule.path] = rule
+
+        self.id_map[rule.id] = rule
+        self.rules.append(rule)
+
+    def load_dict(self, obj: dict, path: Optional[Path] = None):
+        contents = TOMLRuleContents.from_dict(obj)
+        rule = TOMLRule(path=path, contents=contents)
+        self.add_rule(rule)
+
+        return rule
+
+    def load_files(self, paths: Iterable[Path]):
+        """Load multiple files into the collection."""
+        for path in paths:
+            self.load_file(path)
+
+    def _deserialize_toml(self, path: Path) -> dict:
+        # use pytoml instead of toml because of annoying bugs
```

Comment:
```suggestion
        if path in self._toml_load_cache:
            return self._toml_load_cache[path]

        # use pytoml instead of toml because of annoying bugs
```